lib/requestPolicy.ts

```diff
@@ -0,0 +1,56 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+import { HttpRequest } from "./httpRequest";
+import { HttpResponse } from "./httpResponse";
+import { RequestPolicyOptions } from "./requestPolicyOptions";
+import { HttpPipelineLogLevel } from "./httpPipelineLogLevel";
+
+/**
+ * Uses the decorator pattern to add custom behavior when an HTTP request is made.
+ * e.g. add header, user agent, timeout, retry, etc.
+ */
+export interface RequestPolicy {
+    /**
+     * Send the provided HttpRequest request.
+     * @param request The HTTP request to send.
+     * @return A Promise that resolves to the HttpResponse from the targeted server.
+     */
+    send(request: HttpRequest): Promise<HttpResponse>;
+}
+
+/**
+ * A base class implementation of RequestPolicy.
+ */
+export abstract class BaseRequestPolicy implements RequestPolicy {
+    constructor(private _nextPolicy: RequestPolicy, private _options: RequestPolicyOptions) {
```

Comment:
I'd lean toward `public readonly nextPolicy`--getters and setters that encapsulate a field are not really as idiomatic in js

Reply:
I've removed property usages from my implementation. I thought about what you said, and I decided that they were just overhead/wrappers that didn't add any benefit.